libfreerdp/core/rdstls.c

```diff
@@ -0,0 +1,787 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * RDSTLS Security protocol
+ *
+ * Copyright 2023 Joan Torres <joan.torres@suse.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *		 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <freerdp/config.h>
+
+#include <freerdp/log.h>
+#include <freerdp/error.h>
+#include <freerdp/settings.h>
+
+#include <winpr/assert.h>
+#include <winpr/stream.h>
+#include <winpr/wlog.h>
+
+#include "rdstls.h"
+#include "transport.h"
+#include "utils.h"
+
+#define TAG FREERDP_TAG("core.rdstls")
+
+struct rdp_rdstls
+{
+	BOOL server;
+	RDSTLS_STATE state;
+	rdpContext* context;
+	rdpTransport* transport;
+
+	UINT32 resultCode;
+};
+
+/**
+ * Create new RDSTLS state machine.
+ *
+ * @param context A pointer to the rdp context to use
+ *
+ * @return new RDSTLS state machine.
+ */
+
+rdpRdstls* rdstls_new(rdpContext* context, rdpTransport* transport)
+{
+	WINPR_ASSERT(context);
+	WINPR_ASSERT(transport);
+
+	rdpSettings* settings = context->settings;
+	WINPR_ASSERT(settings);
+
+	rdpRdstls* rdstls = (rdpRdstls*)calloc(1, sizeof(rdpRdstls));
+
+	if (!rdstls)
+		return NULL;
+
+	rdstls->context = context;
+	rdstls->transport = transport;
+	rdstls->server = settings->ServerMode;
+
+	return rdstls;
+}
+
+/**
+ * Free RDSTLS state machine.
+ * @param rdstls The RDSTLS instance to free
+ */
+
+void rdstls_free(rdpRdstls* rdstls)
+{
+	if (!rdstls)
+		return;
+
+	free(rdstls);
+}
+
+static const char* rdstls_get_state_str(RDSTLS_STATE state)
+{
+	switch (state)
+	{
+		case RDSTLS_STATE_CAPABILITIES:
+			return "RDSTLS_STATE_CAPABILITIES";
+		case RDSTLS_STATE_AUTH_REQ:
+			return "RDSTLS_STATE_AUTH_REQ";
+		case RDSTLS_STATE_AUTH_RSP:
+			return "RDSTLS_STATE_AUTH_RSP";
+		default:
+			return "UNKNOWN";
+	}
+}
+
+static RDSTLS_STATE rdstls_get_state(rdpRdstls* rdstls)
+{
+	WINPR_ASSERT(rdstls);
+	return rdstls->state;
+}
+
+static BOOL rdstls_set_state(rdpRdstls* rdstls, RDSTLS_STATE state)
+{
+	WINPR_ASSERT(rdstls);
+
+	WLog_DBG(TAG, "-- %s\t--> %s", rdstls_get_state_str(rdstls->state),
```

Comment:
Might be a good idea to check allowed state transitions here and bail out if an invalid one is detected.
Also might be a good idea to check the return of that function where it is used.